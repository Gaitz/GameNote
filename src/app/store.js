import { configureStore } from "@reduxjs/toolkit"
import authenticationReducer from "game-note/features/authentication/authenticationSlice"

const store = configureStore({
    reducer: { authentication: authenticationReducer }
})

export default store
