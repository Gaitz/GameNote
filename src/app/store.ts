import { configureStore } from "@reduxjs/toolkit"
import counterReducer from "game-note/features/example/couter/counterSlice"

const store = configureStore({
  reducer: { counter: counterReducer }
})

export default store
export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
