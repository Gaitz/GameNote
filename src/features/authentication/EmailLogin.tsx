import {
  Button,
  FormControl,
  FormLabel,
  Input,
  Grid,
  InputRightElement,
  GridItem
} from "@chakra-ui/react"
import { useState } from "react"
import { useTranslation } from "next-i18next"
import { useFirebaseAuthEmailLogin } from "game-note/features/authentication"

export function EmailLogin() {
  const { t } = useTranslation("login")

  const EMAIL_FORM_ID = "EmailLogin"
  const EMAIL_INPUT_NAME = "Email"
  const PASSWORD_INPUT_NAME = "Password"

  const {
    handleEmailSignIn,
    handleEmailSignUp,
    isSubmitting
  } = useFirebaseAuthEmailLogin({
    EMAIL_FORM_ID,
    EMAIL_INPUT_NAME,
    PASSWORD_INPUT_NAME
  })

  const [passwordVisible, togglePasswordVisible] = useState(false)

  return (
    <Grid
      gap={2}
      templateColumns={"repeat(2, 1fr)"}
      as="form"
      id={EMAIL_FORM_ID}
      onSubmit={(element) => element.preventDefault()}
    >
      <GridItem colSpan={2}>
        <FormControl isRequired>
          <FormLabel>{t("emailInput")}</FormLabel>
          <Input
            name={EMAIL_INPUT_NAME}
            type="email"
            isRequired
            colorScheme="twitter"
          />
        </FormControl>
      </GridItem>
      <GridItem colSpan={2}>
        <FormControl isRequired gridColumn={2}>
          <FormLabel>{t("passwordInput")}</FormLabel>
          <Input
            autoComplete="current-password"
            name={PASSWORD_INPUT_NAME}
            type={passwordVisible ? "text" : "password"}
            isRequired
            colorScheme="twitter"
          />
          <InputRightElement width="4.5rem">
            <Button
              size="xs"
              onClick={() => togglePasswordVisible(!passwordVisible)}
            >
              {passwordVisible ? t("hide") : t("show")}
            </Button>
          </InputRightElement>
        </FormControl>
      </GridItem>
      <GridItem textAlign="center">
        <Button
          type="submit"
          onClick={handleEmailSignIn}
          isDisabled={isSubmitting}
          colorScheme="purple"
        >
          {t("signIn")}
        </Button>
      </GridItem>
      <GridItem textAlign="center">
        <Button
          type="submit"
          onClick={handleEmailSignUp}
          isDisabled={isSubmitting}
          colorScheme="purple"
        >
          {t("signUp")}
        </Button>
      </GridItem>
    </Grid>
  )
}
