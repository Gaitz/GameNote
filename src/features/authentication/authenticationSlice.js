import { createAsyncThunk, createSlice } from "@reduxjs/toolkit"
import firebaseService from "game-note/services/firebaseService"

const login = createAsyncThunk("authentication/login", async (loginData) => {
  if (loginData.email && loginData.password) {
    const userCredential = await firebaseService.auth.signInWithEmailAndPassword(
      loginData.email,
      loginData.password
    )
    return { email: userCredential.user.email }
  }
})

const authenticationSlice = createSlice({
  name: "authentication",
  initialState: {
    user: null,
  },
  reducers: {},
  extraReducers: {
    [login.fulfilled]: (state, action) => {
      console.log(action)
      state.user = { email: action.payload.email }
    },
  },
})

export { login }
export default authenticationSlice.reducer
