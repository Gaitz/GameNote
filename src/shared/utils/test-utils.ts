import { render } from "@testing-library/react"
import store from "game-note/shared/store"
import theme from "game-note/shared/styles/theme"

// const AllTheProviders = ({ children }) => {
//   return (
//     <ReduxProvider store={store}>
//       <ChakraProvider theme={theme}>{children}</ChakraProvider>
//     </ReduxProvider>
//   )
// }

// const customRender = (ui: ReactElement, options?: object) =>
//   render(ui, { wrapper: AllTheProviders, ...options })

// // re-export everything
// export * from "@testing-library/react"

// // override render method
// export { customRender as render }
