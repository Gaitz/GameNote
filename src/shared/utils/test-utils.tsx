import "@testing-library/jest-dom"
import { render } from "@testing-library/react"
import { Provider as ReduxProvider } from "react-redux"
import { ChakraProvider } from "@chakra-ui/react"
import store from "game-note/shared/store"
import theme from "game-note/shared/styles/theme"
import React, { FC } from "react"
import type { ReactElement } from "react"
import type { RenderOptions } from "@testing-library/react"

/* eslint-disable react/prop-types */
const AllTheProviders: FC = ({ children }) => {
  return (
    <ReduxProvider store={store}>
      <ChakraProvider theme={theme}>{children}</ChakraProvider>
    </ReduxProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from "@testing-library/react"
export { default as userEvent } from "@testing-library/user-event"

export { customRender as render }
