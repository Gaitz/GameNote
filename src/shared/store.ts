import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux"
import { configureStore } from "@reduxjs/toolkit"
import authenticationSlice from "game-note/features/authentication/authenticationSlice"

const store = configureStore({
  reducer: { authentication: authenticationSlice }
})

export default store

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch

export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
