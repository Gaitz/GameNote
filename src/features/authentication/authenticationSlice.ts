import { createSlice, PayloadAction } from "@reduxjs/toolkit"
import firebase from "firebase/app"
import nookies from "nookies"
import { AppThunk } from "game-note/shared/store"

export const AUTH_COOKIE_KEY = "game-note_token"
const SLICE_NAME = "authentication"

interface AuthenticationState {
  user: AppUser
}

export type AppUser = {
    displayName?: string | null,
    email?: string | null,
    photoURL?: string | null
  } | null

const initialState: AuthenticationState = {
  user: null
}

export const authentication = createSlice({
  name: SLICE_NAME,
  initialState,
  reducers: {
    init: () => initialState,
    signIn: (state, action: PayloadAction<AppUser>) => {
      if (action.payload) {
        state.user = action.payload
      }
    },
    signOut: (state) => {
      state.user = null
    }
  }
})

export const logInFromFirebaseAuth = (user: firebase.User): AppThunk => async (dispatch) => {
  if (user) {
    const firebaseToken = await user.getIdToken()
    nookies.set(null, AUTH_COOKIE_KEY, firebaseToken)
    const { displayName, email, photoURL } = user

    dispatch(signIn({
      displayName,
      email,
      photoURL
    }))
  }
}

export const logOutFromFirebaseAuth = (): AppThunk => async (dispatch) => {
  await firebase.auth().signOut()
  nookies.set(null, AUTH_COOKIE_KEY, "")
  dispatch(signOut())
}

export const { signIn, signOut, init } = authentication.actions

export default authentication.reducer

