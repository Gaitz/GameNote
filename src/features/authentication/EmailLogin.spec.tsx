import EmailLogin from "./EmailLogin"
import { render, userEvent, waitFor } from "game-note/shared/utils/test-utils"
import { cleanup } from "@testing-library/react"
import store from "game-note/shared/store"
import firebase from "firebase/app"
import { init } from "./authenticationSlice"

describe("EmailLogin component", () => {
  beforeEach(() => {
    store.dispatch(init())
    // @ts-ignore
    firebase._setAuthSuccess()
  })

  afterEach(() => {
    cleanup()
  })

  it("Snapshot", () => {
    const { asFragment } = render(<EmailLogin />)
    expect(asFragment()).toMatchSnapshot()
  })

  it("Toggle Password Visibility", () => {
    const { getByText, getByTestId } = render(<EmailLogin />)
    const toggleButton = getByText("show")
    const passwordInput = getByTestId("passwordInput")

    expect(passwordInput).toHaveAttribute("type", "password")
    userEvent.click(toggleButton)
    expect(passwordInput).toHaveAttribute("type", "text")
  })

  it("Sign Up", async () => {
    // @ts-ignore
    firebase._setTestMode("signUp")

    const { getByTestId, findByText } = render(<EmailLogin />)
    const emailInput = getByTestId("emailInput")
    const passwordInput = getByTestId("passwordInput")
    const signUp = getByTestId("signUp")

    userEvent.type(emailInput, "test@test.test")
    userEvent.type(passwordInput, "test_password")
    await waitFor(() => userEvent.click(signUp))

    expect(store.getState()["authentication"]).toMatchInlineSnapshot(`
      Object {
        "user": Object {
          "displayName": "test_signUp_displayName",
          "email": "test_signUp_email",
          "photoURL": "test_signUp_photoURL",
        },
      }
    `)

    // @ts-ignore
    firebase._setAuthSuccess(false)

    await waitFor(() => userEvent.click(signUp))

    const errorToast = await findByText("test_signUp_errorCode")
    expect(errorToast).toBeInTheDocument()
  })

  it("Sign In", async () => {
    // @ts-ignore
    firebase._setTestMode("signIn")

    const { getByTestId, findByText } = render(<EmailLogin />)
    const emailInput = getByTestId("emailInput")
    const passwordInput = getByTestId("passwordInput")
    const signIn = getByTestId("signIn")

    userEvent.type(emailInput, "test@test.test")
    userEvent.type(passwordInput, "test_password")
    await waitFor(() => userEvent.click(signIn))

    expect(store.getState()["authentication"]).toMatchInlineSnapshot(`
      Object {
        "user": Object {
          "displayName": "test_signIn_displayName",
          "email": "test_signIn_email",
          "photoURL": "test_signIn_photoURL",
        },
      }
    `)

    // @ts-ignore
    firebase._setAuthSuccess(false)

    await waitFor(() => userEvent.click(signIn))

    const errorToast = await findByText("test_signIn_errorCode")
    expect(errorToast).toBeInTheDocument()
  })
})
