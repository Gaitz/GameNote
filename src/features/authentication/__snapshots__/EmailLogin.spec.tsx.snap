// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailLogin component Snapshot 1`] = `
<DocumentFragment>
  <form
    class="css-vgka4c"
    id="EmailLogin"
  >
    <div
      class="css-tuh9u2"
    >
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-10m1y3x"
          for="field-1"
          id="field-1-label"
        >
          emailInput
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-1tknnrp"
            role="presentation"
          >
            *
          </span>
        </label>
        <input
          aria-required="true"
          autocomplete="username"
          class="chakra-input css-1b73kmm"
          data-testid="emailInput"
          id="field-1"
          name="Email"
          required=""
          type="email"
        />
      </div>
    </div>
    <div
      class="css-tuh9u2"
    >
      <div
        class="chakra-form-control css-1goklfu"
        role="group"
      >
        <label
          class="chakra-form__label css-10m1y3x"
          for="Password"
          id="field-2-label"
        >
          passwordInput
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-1tknnrp"
            role="presentation"
          >
            *
          </span>
        </label>
        <input
          aria-required="true"
          autocomplete="current-password"
          class="chakra-input css-1b73kmm"
          data-testid="passwordInput"
          id="field-2"
          name="Password"
          required=""
          type="password"
        />
        <div
          class="chakra-input__right-element css-1tbxs5h"
        >
          <button
            class="chakra-button css-10z31di"
            data-testid="toggleVisible"
            type="button"
          >
            show
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-xi606m"
    >
      <button
        class="chakra-button css-16trzbp"
        data-testid="signIn"
        type="submit"
      >
        signIn
      </button>
    </div>
    <div
      class="css-xi606m"
    >
      <button
        class="chakra-button css-16trzbp"
        data-testid="signUp"
        type="submit"
      >
        signUp
      </button>
    </div>
  </form>
</DocumentFragment>
`;
